modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/StatisticsProcessor.java

```diff
@@ -0,0 +1,615 @@
+public class StatisticsProcessor {
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /** Ignite statistics repository. */
+    private final IgniteStatisticsRepository statRepo;
+
+    /** Ignite Thread pool executor to do statistics collection tasks. */
+    private final IgniteThreadPoolExecutor gatherPool;
+
+    /** (cacheGroupId -> gather context) */
+    private final ConcurrentMap<StatisticsKey, LocalStatisticsGatheringContext> gatheringInProgress =
+        new ConcurrentHashMap<>();
+
+    /** Active flag (used to skip commands in inactive cluster.) */
+    private volatile boolean active;
+
+    /* Lock protection of started gathering during deactivation. */
+    private static final GridBusyLock busyLock = new GridBusyLock();
+
+    /**
+     * Constructor.
+     *
+     * @param repo IgniteStatisticsRepository.
+     * @param gatherPool Thread pool to gather statistics in.
+     * @param logSupplier Log supplier function.
+     */
+    public StatisticsProcessor(
+        IgniteStatisticsRepository repo,
+        IgniteThreadPoolExecutor gatherPool,
+        Function<Class<?>, IgniteLogger> logSupplier
+    ) {
+        this.statRepo = repo;
+        this.gatherPool = gatherPool;
+        this.log = logSupplier.apply(StatisticsProcessor.class);
+    }
+
+    /**
+     * Update statistics for the given key to actual state.
+     * If byObsolescence and tbl is not {@code null} - does not clear any other partitions.
+     * Should run throw management pool only.
+     * 1) Replace previous gathering context if exist and needed (replace byObsolescence gathering with new one or
+     * replace gathering with older configuration or topology version with new one).
+     * 2) If byObsolescence and no table awailable - clean obsolescence and partition statistics for the given key.
+     * 3) Submit tasks for each specified partition.
+     * 4) after last task finish gathering - it starts aggregation.
+     * 5) read all partitions & obsolescence from repo and
+     * if byObsolescence = {@code true} - remove unnecessary one and aggregate by specified list
+     * if byObsolescence = {@code false} - aggregate all presented in store (because it should contains only actual ones)
+     * 5) save aggregated local statistics
+     *
+     * @param byObsolescence Update only obsolescence partitions.
+     * @param tbl Table to update. If {@code null} - just clear all partitions and obsolescence from the repo
+     * @param cfg Statistics configuration to use.
+     * @param partsToProcess Partitions to update, if !byObsolescence - all primary partitions for the given topology.
+     * @param topVer Topology version, can be {@code null} if tbl is null.
+     */
+    public void updateKeyAsync(
+        boolean byObsolescence,
+        GridH2Table tbl,
+        StatisticsObjectConfiguration cfg,
+        Set<Integer> partsToProcess,
+        AffinityTopologyVersion topVer
+    ) {
+        if (!startJob("Updating key " + cfg.key()))
```

Comment:
String is calculated every time, but logged only in debug.

Reply:
Removed.